crates/ruff_db/src/vendored.rs

```diff
@@ -0,0 +1,379 @@
+    /// The path exists in the zip archive and represents a vendored file
+    File,
+
+    /// The path exists in the zip archive and represents a vendored directory of files
+    Directory,
+}
+
+#[derive(Debug)]
+struct VendoredFileSystemInner(Mutex<RefCell<VendoredZipArchive>>);
+
+impl VendoredFileSystemInner {
+    fn new(raw_bytes: &'static [u8]) -> Result<Self> {
+        Ok(Self(Mutex::new(RefCell::new(VendoredZipArchive::new(
+            raw_bytes,
+        )?))))
+    }
+
+    /// Acquire a lock on the underlying zip archive.
+    /// The call will block until it is able to acquire the lock.
+    ///
+    /// ## Panics:
+    /// If the current thread already holds the lock.
+    fn lock(&self) -> LockedZipArchive {
+        LockedZipArchive(self.0.lock().unwrap())
+    }
+}
+
+#[derive(Debug)]
+struct LockedZipArchive<'a>(MutexGuard<'a, RefCell<VendoredZipArchive>>);
+
+impl<'a> LockedZipArchive<'a> {
+    fn borrow_mut(&self) -> ArchiveReader {
+        ArchiveReader(self.0.borrow_mut())
+    }
+}
+
+#[derive(Debug)]
+struct ArchiveReader<'a>(RefMut<'a, VendoredZipArchive>);
+
+impl<'a> Deref for ArchiveReader<'a> {
+    type Target = VendoredZipArchive;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl<'a> DerefMut for ArchiveReader<'a> {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.0
+    }
+}
+
+/// Newtype wrapper around a ZipArchive.
+#[derive(Debug)]
+struct VendoredZipArchive(ZipArchive<io::Cursor<&'static [u8]>>);
+
+impl VendoredZipArchive {
+    pub fn new(data: &'static [u8]) -> Result<Self> {
+        Ok(Self(ZipArchive::new(io::Cursor::new(data))?))
+    }
+
+    fn lookup_path(&mut self, path: &NormalizedVendoredPath) -> Result<ZipFile> {
+        Ok(self.0.by_name(path.as_str())?)
+    }
+}
+
+/// A path that has been normalized via the `normalize_vendored_path` function.
+///
+/// Trailing slashes are normalized away by `camino::Utf8PathBuf`s,
+/// but trailing slashes are crucial for distinguishing between
+/// files and directories inside zip archives.
+#[derive(Debug, Clone, PartialEq, Eq)]
+struct NormalizedVendoredPath(String);
+
+impl NormalizedVendoredPath {
+    fn with_trailing_slash(mut self) -> Self {
+        debug_assert!(!self.0.ends_with('/'));
+        self.0.push('/');
+        self
+    }
+
+    fn as_str(&self) -> &str {
+        self.0.as_str()
+    }
+}
+
+/// Normalizes the path by removing `.` and `..` components.
+///
+/// ## Panics:
+/// If a path with an unsupported component for vendored paths is passed.
+/// Unsupported components are path prefixes,
+/// and path root directories appearing anywhere except at the start of the path.
+fn normalize_vendored_path(path: &VendoredPath) -> NormalizedVendoredPath {
+    debug_assert!(
+        !path.as_std_path().is_absolute(),
+        "Attempted to lookup an absolute filesystem path relative to a vendored zip archive"
+    );
+    let mut normalized_parts = camino::Utf8PathBuf::new();
```

Comment:
We don't want to use `is_absolute`. It's platform dependent, meaning `VendoredPath`'s suddenly become platform specific. 

If you want to use linux semantics, use `!path.starts_with("/")`

Reply:
It's very surprising to me that `is_absolute` is platform specific. Thanks for the reminder!

Reply:
Oh wait, no, you said "platform-dependent", not "platforms-specific". I think "platform-dependent" is what I want here. Whatever kind of path we're dealing with here, it should never have been resolved to an absolute OS-filesystem path. It should always be a path that's relative to the root of the zip directory. If this assertion ever fails on either Linux or Windows, I think that indicates a logical error in our code somewhere.

Reply:
A path starting with `/` will fail on Windows.

Reply:
I'm still not sure I fully understand. The main purpose of this debug assertion is to make sure that the path hasn't been `canonicalize`d or otherwise "made absolute" prior to attempting to resolve the path relative to the zip archive. On a Unix system `canonicalize()` would mean that the path ould start with `/`, but on a Windows system it would mean that it would start with a prefix, right? So I think `is_absolute()` is what we want here?

Reply:
I think we want `VendoredPath` to be file system agnostic. It's not a real file system so we should avoid the madness of having different path representations for different systems. I'm also not sure what happens if you lookup the path `C:\test` in the zip file. I would simply assume that this would outright fail?

I recommend that `VendoredPath` uses unix representation exclusively.

Reply:
`VendoredPath` does use unix representation exclusively; this PR doesn't change that, and nor does this assertion. I think my takeaway here is that this debug assertion causes more confusion than clarification, so I think I'll just remove it altogether :)

Reply:
The debug assertion itself is fine, but it can't be `is_absolute` because `is_absolute` returns `false` on Windows if the path starts with `/`

Reply:
I understand that. I still think we're talking past each other a little bit, but I've got rid of the assertion now anyway.